# Puzzle

https://thefiddler.substack.com/p/can-your-team-self-organize

**This Week’s Fiddler**

...

Suppose a total of four people are participating in this team building exercise. They all write down different numbers, and then independently place their names at random positions on the line.

On average, what do you expect the team’s score to be?

**This Week’s Extra Credit**

Instead of four people, now suppose there are 10 people participating in the team building exercise. As before, they all write down different numbers, and then independently place their names at random positions on the line.

On average, what do you expect the team’s score to be?


# Solution

I worked out the fiddler case mentally, and I believe the answer is 29/12 = 2.41667

Lacking grand insights, I am going to write some code for the extra credit.

In [27]:
# The copilot suggested code is tidy, but there is no reuse of computed results.
# The recursive code is less tidy, but it does reuse computed results.
def find_longest_increasing_subseq_length(arr):
    if not arr:
        return 0
    n = len(arr)
    lis = [1] * n  # Initialize LIS values for all indexes as 1
    # Compute optimized LIS values in a bottom-up manner
    for i in range(1, n):
        for j in range(i):
            if arr[i] > arr[j] and lis[i] < lis[j] + 1:
                lis[i] = lis[j] + 1
    return max(lis)

assert find_longest_increasing_subseq_length('ACBD') == 3

from functools import cache
@cache
def find_longest_increasing_subseq_length_recursive(arr):
    length = len(arr)
    result = 0
    if length <= 1:
        result = length
    elif length == 2:
        result = 2 if arr[1] > arr[0] else 1
    else:
        max_pos = 0
        max_elem = arr[0]
        for i in range(1,length):
            if arr[i] > max_elem:
                max_elem = arr[i]
                max_pos = i
        maxlen_incl_maxelem, maxlen_excl_maxelem = 0, 0
        if max_pos > 0:
            maxlen_incl_maxelem = 1 + find_longest_increasing_subseq_length_recursive(arr[:max_pos])
        if max_pos < length - 1:
            maxlen_excl_maxelem = find_longest_increasing_subseq_length_recursive(arr[:max_pos]+arr[max_pos + 1:])
        result = max(maxlen_incl_maxelem, maxlen_excl_maxelem)
    return result

assert find_longest_increasing_subseq_length_recursive('ACBDGEHF') == 5

In [28]:
from itertools import permutations
def expected_lislen(arr, lisl_fn=find_longest_increasing_subseq_length):
    E_sum = 0
    E_cnt = 0
    for perm in permutations(arr):
        E_sum += lisl_fn(perm)
        E_cnt += 1
    return E_sum / E_cnt

assert expected_lislen('ABC', find_longest_increasing_subseq_length) == 2
assert expected_lislen('ABC', find_longest_increasing_subseq_length_recursive) == 2

In [29]:
Fiddler_answer = expected_lislen('ABCD', find_longest_increasing_subseq_length)
print(f'Fiddler_answer={Fiddler_answer}')
Fiddler_answer = expected_lislen('ABCD', find_longest_increasing_subseq_length_recursive)
print(f'Fiddler_answer={Fiddler_answer}')

Fiddler_answer=2.4166666666666665
Fiddler_answer=2.4166666666666665


In [30]:
Extra_Credit_answer = expected_lislen('ABCDEFGHIJ', find_longest_increasing_subseq_length)
print(f'Extra_Credit_answer={Extra_Credit_answer}')
Extra_Credit_answer = expected_lislen('ABCDEFGHIJ', find_longest_increasing_subseq_length_recursive)
print(f'Extra_Credit_answer={Extra_Credit_answer}')

Extra_Credit_answer=4.33496059303351
Extra_Credit_answer=4.33496059303351


In [32]:
for i in range(2, 12):
    arr = ''.join(chr(ord('A') + j) for j in range(i))
    ans = expected_lislen(arr, find_longest_increasing_subseq_length_recursive)
    print(f'arr={arr}, expected_lislen={ans}')

arr=AB, expected_lislen=1.5
arr=ABC, expected_lislen=2.0
arr=ABCD, expected_lislen=2.4166666666666665
arr=ABCDE, expected_lislen=2.7916666666666665
arr=ABCDEF, expected_lislen=3.140277777777778
arr=ABCDEFG, expected_lislen=3.4652777777777777
arr=ABCDEFGH, expected_lislen=3.7703373015873014
arr=ABCDEFGHI, expected_lislen=4.059350198412698
arr=ABCDEFGHIJ, expected_lislen=4.33496059303351
arr=ABCDEFGHIJK, expected_lislen=4.598861055996473


# Conclusions

**Fiddler_answer=2.4166666666666665**

**Extra_Credit_answer=4.33496059303351**

The recursive code is faster, but not by that much at n=10.

The runtime seems to increase quickly beyond n=10, but that might be an artifact of the cache.